torchbenchmark/util/distributed/deepspeed.py

```diff
@@ -0,0 +1,66 @@
+from .trainer import Trainer
+from torch.nn.parallel import DistributedDataParallel as DDP
+from torchbenchmark.util.e2emodel import E2EBenchmarkModel
+import torch.distributed as dist
+
+class DeepSpeedTrainer(Trainer):
+    DEFAULT_MEASURE_ITERATIONS = 10
+    def __init__(self, args, model_class, batch_size=None, extra_args=[]):
+        super().__init__(args, model_class, mode="SPMD")
+        
+        self.setup()
+        # create model instance after Trainer setup, so that
+        # visible devices won't be revised in model constructor
+
+        #TODO(whc) use_deepspeed breaks the API... but it is most convenient to initialize/configure deepspeed
+        # inside the model constructur rather than later, since later is too late to intercept the Accelerator
+        # configuration for hf_ models
+        model: E2EBenchmarkModel = model_class("train", batch_size, use_deepspeed=True, extra_args=extra_args)
```

Comment:
The changes to DDPTrainer and the new DeepSpeedTrainer looks OK to me. But I agree that this is a bit inconsistent where DDPTrainer explicitly creates DDP instance, while DeepSpeedTrainer relies on `hf_bert` model to do that. 

Curious, (not saying this need to be changed), is there any blocker that prevents us from 1) creating a regular hf_bert model without DeepSpeed on this line 2) use the following API to wrap it with DeepSpeed?

https://github.com/microsoft/DeepSpeedExamples/blob/89ce0d5a40cec1772c8673bf01c532828ce3d31d/cifar/cifar10_deepspeed.py#L246-L247

I am not familiar with DeepSpeed APIs. There might be gaps that I am not aware. 

cc @zhaojuanmao

Reply:
> Curious, (not saying this need to be changed), is there any blocker that prevents us from 1) creating a regular hf_bert model without DeepSpeed on this line 2) use the following API to wrap it with DeepSpeed?

i think that is totally possible.  However, I wanted to err on the side of using the 'native' integration with deepspeed rather than baking our own, just in case there are some differences.


> this is a bit inconsistent where DDPTrainer explicitly creates DDP instance, while DeepSpeedTrainer relies on hf_bert model to do that.

it would be pretty easy to move the ddp thing inside hf_bert to make it consistent.  But since we aren't adding any real configuration for DDP, it seems like having it done 'in common' across models is feasible, and so why not keep it there even at the cost of being inconsistent with deepspeed?